runtime/hashing/common/src/aws/smithy/kotlin/runtime/hashing/HashingSink.kt

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.hashing
+
+import aws.smithy.kotlin.runtime.io.SdkSink
+import aws.smithy.kotlin.runtime.io.internal.SdkSinkObserver
+
+/**
+ * A sink which hashes data as it is being written to
+ * @param hash The [HashFunction] to hash data with
+ * @param sink the [SdkSink] to hash
+ */
+public class HashingSink(public val hash: HashFunction, sink: SdkSink) : HashFunction by hash, SdkSinkObserver(sink) {
```

Comment:
**Comment**: Making the underlying `hash` publicly visible is inadvisable since that means its state can be mutated by object holders. We should probably make it private instead.

Reply:
Or use a `HashSupplier` instead.

Reply:
The reason I made this public was for the `digestValue()` use case for CRC hash functions. I'm not sure how to nicely provide this function if the underlying hash is private, but I understand the problem with mutation by object holders.

From what I see, HashSupplier wouldn't be great here because it returns a new HashFunction each invocation, while we want to use the same HashFunction object to get the hash value of the entire underlying Source/Sink.

For `digestValue()`, I could do this:
```kotlin
public class HashingSource(private val hash: HashFunction, source: SdkSource) : SdkSourceObserver(source) {
    override fun observe(data: ByteArray, offset: Int, length: Int) {
        hash.update(data, offset, length)
    }

    public fun digestValue() : UInt? = (hash as Crc32?)?.digestValue()
}
```

But it seems too special-cased for a generic HashingSource

Reply:
Yes, a single unsigned-int digest value is very specific to CRC32. We shouldn't include it in anything that utilizes generic hash functionality. How will this be used eventually outside of things like tests?

Reply:
It's being used for event steams today. I'm adding it only to change event streams' [CrcSource implementation](https://github.com/awslabs/aws-sdk-kotlin/blob/c099049829d6a45307173383c50bf40f0095825d/aws-runtime/protocols/aws-event-stream/common/src/aws/sdk/kotlin/runtime/protocol/eventstream/CrcUtil.kt#L18) to this new one.

Reply:
Why does the event streams code need to treat the CRC as a `UInt`? Could it just as easily use byte arrays?

Reply:
It's possible. Event streams sends the expected checksum value as `UInt`, so we need to either
1. compare using digestValue()'s `UInt` (how it's done today)
2. convert the given expected checksum from `UInt` to `ByteArray`, and compare using `digest()` instead

I didn't want to make too many changes in event streams besides replacing the CrcSource/Sink that was used there, which is why I went with option 1.

Reply:
Discussed offline, we decided to use just `digest()`/`ByteArray` everywhere for hash functions _including_ CRC32. We'll remove specialized `digestValue()` members for the source/sink.